### ISSUE TITLE
AdamOptimizer gradient path uses Vector flattening, incompatible with Transformer output shapes

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1115]
## Bug Description

After fixing #1113 (feature selection) and #1114 (loss function shape mismatch), Transformer training through `AiModelBuilder` crashes in the gradient computation path. `ComputeGradients` in `NeuralNetworkBase` calls `MeanSquaredErrorLoss.CalculateDerivative` which expects `Vector<T>` — but the Transformer's output tensor flattened to a vector has a different length than the target vector.

## Error

```
System.ArgumentException: Predicted and actual vectors must have the same length.
  at LossFunctionBase.ValidateVectorLengths()
  at MeanSquaredErrorLoss.CalculateDerivative()
  at NeuralNetworkBase.ComputeGradients()
  at GradientBasedOptimizerBase.CalculateGradient()
  at AdamOptimizer.Optimize()
```

## Root Cause

The `AdamOptimizer` gradient path (`GradientBasedOptimizerBase.CalculateGradient`) calls `NeuralNetworkBase.ComputeGradients` which:
1. Predicts with the input → gets output tensor `[batch, seq_len, vocab_size]` (e.g., `[11, 4, 38]`)
2. Flattens to a `Vector<T>` of length 1672
3. The target tensor `[11, 4]` flattens to length 44
4. `MeanSquaredErrorLoss.CalculateDerivative(predicted: len=1672, actual: len=44)` → crash

The optimizer's gradient computation pipeline assumes the model output and target have the same shape when flattened. For Transformers doing next-token prediction, the output has an extra vocabulary dimension.

Note: The `ComputeTapeLoss` path (fixed in #1114) handles this correctly via one-hot encoding. But the `CalculateDerivative` path in the traditional optimizer loop does not.

## Suggested Fix

The `ComputeGradients` or `CalculateGradient` method should use the tape-based loss when the model is a neural network, instead of flattening to vectors. Alternatively, the Adam optimizer could detect when output shapes differ and either:
1. One-hot encode the target to match, or
2. Use `TrainWithTape` (which already works) instead of the legacy vector-based gradient path

## Related Issues

- #1113: Feature selection crashes for Transformers (fixed)
- #1114: CategoricalCrossEntropyLoss shape mismatch (fixed for tape path)
- This issue: The legacy vector-based gradient path also has shape mismatches

Together, these three issues block Transformer training through `AiModelBuilder`.
